## Try to read txt file and separated by space

In [1]:
with open("./file1.txt") as f:
    for line in f:
        print(line.replace("\n", "").split(' '))

['01', '2']
['02', '2']
['03', '4']
['04', '4']
['05', '12']
['06', '12']


## Let's construct dictionary to save table info

In [2]:
table_info = {}
table_info["Table Number List"] = []
table_info["Amount Limit"] = {}
with open("./file1.txt") as f:
    for line in f:
        split_list = line.replace("\n", "").split(' ')
        table_no = split_list[0]
        amount_limit = split_list[1]
        table_info["Table Number List"].append(table_no)
        table_info["Amount Limit"][table_no] = int(amount_limit)
table_info

{'Table Number List': ['01', '02', '03', '04', '05', '06'],
 'Amount Limit': {'01': 2, '02': 2, '03': 4, '04': 4, '05': 12, '06': 12}}

---

### Step 1. Load tables information from a file

Below is the program output for loading data from the input file (Bold red texts are user input):

---

Table file:

<span style="color:red">file1.txt</span>

Imported 6 table(s).

---

In [3]:
def construct_table_info_dict(_file_path: str):
    table_info = {}
    table_info["Table Number List"] = []
    table_info["Amount Limit"] = {}
    with open(_file_path) as f:
        for line in f:
            split_list = line.replace("\n", "").split(' ')
            table_no = split_list[0]
            amount_limit = split_list[1]
            table_info["Table Number List"].append(table_no)
            table_info["Amount Limit"][table_no] = int(amount_limit)
    return table_info

In [4]:
print("Table file:")
file_path = input()
table_info = construct_table_info_dict(_file_path=file_path)
print(f"Imported {len(table_info['Table Number List'])} table(s).")

Table file:
file1.txt
Imported 6 table(s).


---

### Step 2. Exit – terminate the program

If the user enters “Exit”, the program prints “Bye” and then terminates.

---

Table file:

<span style="color:red">file1.txt</span>

Imported 6 table(s).

<span style="color:red">Exit</span>

Bye

---

In [5]:
print("Table file:")
file_path = input()
table_info = construct_table_info_dict(_file_path=file_path)
print(f"Imported {len(table_info['Table Number List'])} table(s).")

command = input()
if command.upper() == "EXIT":
    print("Bye")

Table file:
file1.txt
Imported 6 table(s).
Exit
Bye


---

### Add some error handling

In [6]:
import os

In [7]:
print("Table file:")
file_path = input()

if not os.path.exists(file_path):
    print(f"{file_path} does NOT exist!")
    raise SystemExit

table_info = construct_table_info_dict(_file_path=file_path)
print(f"Imported {len(table_info['Table Number List'])} table(s).")

close_system = False
while not close_system:
    command = input()
    if command.upper() == "EXIT":
        print("Bye")
        close_system = True
    print("Wrong command, please enter 'Exit' to close system!")

Table file:
file1.txt
Imported 6 table(s).
Book|CK Lai|91234567|01-04-2022|10
Wrong command, please enter 'Exit' to close system!
Exit
Bye
Wrong command, please enter 'Exit' to close system!


---

### Step 3. Book – request a booking

In the test case below, 3 bookings are added to the system.

---

Table file:

<span style="color:red">file1.txt</span>

Imported 6 table(s).

<span style="color:red">Book|CK Lai|91234567|01-04-2022|10</span>

Added booking. The ticket code for 01-04-2022 is 1.

<span style="color:red">Book|Tom|61234567|01-04-2022|3</span>

Added booking. The ticket code for 01-04-2022 is 2.

<span style="color:red">Book|Jolly|51234567|02-04-2022|5</span>

Added booking. The ticket code for 02-04-2022 is 1.

<span style="color:red">Exit</span>

Bye

---

In [8]:
import json 

In [10]:
BookingSystemData = {}
BookingSystemData["Table Info"] = table_info
BookingSystemData["Booking Queue"] = {}
BookingSystemData["Booking Queue"]["Date Queue List"] = {}

print("Enter booking information:")
client_booking_info = input()
client_info_split_list = client_booking_info.split("|")
if client_info_split_list[0] == "Book":
    client_name = client_info_split_list[1]
    phone_number = client_info_split_list[2]
    booking_date = client_info_split_list[3]
    booking_amount = int(client_info_split_list[4])
    if booking_date not in BookingSystemData["Booking Queue"]["Date Queue List"]:
        BookingSystemData["Booking Queue"]["Date Queue List"][booking_date] = [client_name]
    else:
        BookingSystemData["Booking Queue"]["Date Queue List"][booking_date].append(client_name)
    if client_name not in BookingSystemData["Booking Queue"]:
        BookingSystemData["Booking Queue"][client_name] = {
            "Phone": phone_number,
            "Booking Date List": [booking_date],
            "Booking Amount": {booking_date: booking_amount}
        }
    else:
        BookingSystemData["Booking Queue"][client_name]["Booking Date List"].append(booking_date)
        BookingSystemData["Booking Queue"][client_name]["Booking Amount"][booking_date] = booking_amount
    queue_no = BookingSystemData["Booking Queue"]["Date Queue List"][booking_date].index(client_name)+1
    print(f"Added booking. The ticket code for {booking_date} is {queue_no}.")
print("\n===\n")
print(json.dumps(BookingSystemData, indent=4))

Enter booking information:
Book|CK Lai|91234567|01-04-2022|10
Added booking. The ticket code for 01-04-2022 is 1.

===

{
    "Table Info": {
        "Table Number List": [
            "01",
            "02",
            "03",
            "04",
            "05",
            "06"
        ],
        "Amount Limit": {
            "01": 2,
            "02": 2,
            "03": 4,
            "04": 4,
            "05": 12,
            "06": 12
        }
    },
    "Booking Queue": {
        "Date Queue List": {
            "01-04-2022": [
                "CK Lai"
            ]
        },
        "CK Lai": {
            "Phone": "91234567",
            "Booking Date List": [
                "01-04-2022"
            ],
            "Booking Amount": {
                "01-04-2022": 10
            }
        }
    }
}


In [11]:
def add_booking_info(
    _booking_system_data: dict,
    _client_booking_info: str
):
    if "Booking Queue" not in _booking_system_data:
        _booking_system_data["Booking Queue"] = {}
        _booking_system_data["Booking Queue"]["Date Queue List"] = {}

    client_info_split_list = _client_booking_info.split("|")
    
    if client_info_split_list[0] == "Book":
        client_name = client_info_split_list[1]
        phone_number = client_info_split_list[2]
        booking_date = client_info_split_list[3]
        booking_amount = int(client_info_split_list[4])
    
        if booking_date not in _booking_system_data["Booking Queue"]["Date Queue List"]:
            _booking_system_data["Booking Queue"]["Date Queue List"][booking_date] = [client_name]
        else:
            _booking_system_data["Booking Queue"]["Date Queue List"][booking_date].append(client_name)
    
        if client_name not in _booking_system_data["Booking Queue"]:
            _booking_system_data["Booking Queue"][client_name] = {
                "Phone": phone_number,
                "Booking Date List": [booking_date],
                "Booking Amount": {booking_date: booking_amount}
            }
            print()
        else:
            _booking_system_data["Booking Queue"][client_name]["Booking Date List"].append(booking_date)
            _booking_system_data["Booking Queue"][client_name]["Booking Amount"][booking_date] = booking_amount
        queue_no = _booking_system_data["Booking Queue"]["Date Queue List"][booking_date].index(client_name)+1
        print(f"Added booking. The ticket code for {booking_date} is {queue_no}.")

In [12]:
import os
import json

In [13]:
print("Table file:")
file_path = input()

if not os.path.exists(file_path):
    print(f"{file_path} does NOT exist!")
    raise SystemExit

table_info = construct_table_info_dict(_file_path=file_path)
print(f"Imported {len(table_info['Table Number List'])} table(s).")

BookingSystemData = {}
BookingSystemData["Table Info"] = table_info

close_system = False
while not close_system:
    print("Enter booking information:") # Can be removed
    client_booking_info = input()
    
    if client_booking_info.upper() == "EXIT":
        print("Bye")
        close_system = True
    else:
        add_booking_info(
            _booking_system_data=BookingSystemData,
            _client_booking_info=client_booking_info
        )

Table file:
file1.txt
Imported 6 table(s).
Enter booking information:
Book|CK Lai|91234567|01-04-2022|10

Added booking. The ticket code for 01-04-2022 is 1.
Enter booking information:
Book|Tom|61234567|01-04-2022|3

Added booking. The ticket code for 01-04-2022 is 2.
Enter booking information:
Book|Jolly|51234567|02-04-2022|5

Added booking. The ticket code for 02-04-2022 is 1.
Enter booking information:
Exit
Bye


In [14]:
print(json.dumps(BookingSystemData, indent=4))

{
    "Table Info": {
        "Table Number List": [
            "01",
            "02",
            "03",
            "04",
            "05",
            "06"
        ],
        "Amount Limit": {
            "01": 2,
            "02": 2,
            "03": 4,
            "04": 4,
            "05": 12,
            "06": 12
        }
    },
    "Booking Queue": {
        "Date Queue List": {
            "01-04-2022": [
                "CK Lai",
                "Tom"
            ],
            "02-04-2022": [
                "Jolly"
            ]
        },
        "CK Lai": {
            "Phone": "91234567",
            "Booking Date List": [
                "01-04-2022"
            ],
            "Booking Amount": {
                "01-04-2022": 10
            }
        },
        "Tom": {
            "Phone": "61234567",
            "Booking Date List": [
                "01-04-2022"
            ],
            "Booking Amount": {
                "01-04-2022": 3
            }
        },

---

### Step 4. ListBookings

Upon completion, the system goes back to the main menu.

---

Table file:

<span style="color:red">file1.txt</span>

Imported 6 table(s).

<span style="color:red">Book|HF|81234567|01-04-2022|5</span>

Added booking. The ticket code for 01-04-2022 is 1.

<span style="color:red">Book|Liu|21234567|01-04-2022|2</span>

Added booking. The ticket code for 01-04-2022 is 2.

<span style="color:red">ListBookings</span>

Booking(s):

HF, 81234567, 01-04-2022 (Ticket 1), 5, Pending.

Liu, 21234567, 01-04-2022 (Ticket 2), 2, Pending.

<span style="color:red">Exit</span>

Bye

---

In [15]:
BookingSystemData["Booking Queue"].keys()

dict_keys(['Date Queue List', 'CK Lai', 'Tom', 'Jolly'])

In [16]:
booking_client_list = list(BookingSystemData["Booking Queue"].keys())
booking_client_list.remove("Date Queue List")

# Get client's name
client_name = booking_client_list[0]
print("Name:", client_name)

# Get client's phone
print("Phone:", BookingSystemData["Booking Queue"][client_name]["Phone"])

# Get client's booking date
print("All booking dates:", BookingSystemData["Booking Queue"][client_name]["Booking Date List"])
booking_date = BookingSystemData["Booking Queue"][client_name]["Booking Date List"][0]
print("Booking date:", booking_date)

# Get client's queue number
print("Queue number:", BookingSystemData["Booking Queue"]["Date Queue List"][booking_date].index(client_name)+1)

# Get client's booking amount
print("Booking amount:", BookingSystemData["Booking Queue"][client_name]["Booking Amount"][booking_date])

Name: CK Lai
Phone: 91234567
All booking dates: ['01-04-2022']
Booking date: 01-04-2022
Queue number: 1
Booking amount: 10


In [17]:
def list_bookings(_booking_sysem_data: dict):
    
    if "Booking Queue" not in _booking_sysem_data:
        print("No booking.")
        return
    
    if len(_booking_sysem_data["Booking Queue"]) < 2:
        print("No booking.")
        return
    
    booking_client_list = list(_booking_sysem_data["Booking Queue"].keys())
    booking_client_list.remove("Date Queue List")

    # Get client's name
    for i in range(len(booking_client_list)):
        client_name = booking_client_list[i]

        # Get client's phone
        phone = _booking_sysem_data["Booking Queue"][client_name]["Phone"]

        # Get client's booking date
        all_booking_date_list = _booking_sysem_data["Booking Queue"][client_name]["Booking Date List"]
        booking_date = all_booking_date_list[0]

        # Get client's queue number
        queue_number = _booking_sysem_data["Booking Queue"]["Date Queue List"][booking_date].index(client_name)+1

        # Get client's booking amount
        booking_amount = _booking_sysem_data["Booking Queue"][client_name]["Booking Amount"][booking_date]

        print(f"{client_name}, {phone}, {booking_date} (Ticket {queue_number}), {booking_amount}, Pending.")

In [18]:
print("Table file:")
file_path = input()

if not os.path.exists(file_path):
    print(f"{file_path} does NOT exist!")
    raise SystemExit

table_info = construct_table_info_dict(_file_path=file_path)
print(f"Imported {len(table_info['Table Number List'])} table(s).")

BookingSystemData = {}
BookingSystemData["Table Info"] = table_info

close_system = False
while not close_system:
    command = input()
    
    if command.upper() == "EXIT":
        print("Bye")
        close_system = True
    elif "BOOK|" in command.upper():
        add_booking_info(
            _booking_system_data=BookingSystemData,
            _client_booking_info=command
        )
    elif command.lower() == "listbookings":
        list_bookings(_booking_sysem_data=BookingSystemData)

Table file:
file1.txt
Imported 6 table(s).
ListBookings
No booking.
Book|HF|81234567|01-04-2022|5

Added booking. The ticket code for 01-04-2022 is 1.
Book|Liu|21234567|01-04-2022|2

Added booking. The ticket code for 01-04-2022 is 2.
ListBookings
HF, 81234567, 01-04-2022 (Ticket 1), 5, Pending.
Liu, 21234567, 01-04-2022 (Ticket 2), 2, Pending.
Exit
Bye


## TBA